### ISSUE TITLE
Duplicate `insertId` when running `executeAsync` in a transaction

### ISSUE THREAD

**Issue by craftzdog** [OP-Engineering/op-sqlite#111]
Hi! I think I found a bug regarding transactions.

When running `executeAsync` to insert rows simultaneously, it randomly gets duplicate `insertId`.

**Versions:**

- OS and version: iOS 17.5 simulator and Android API 34 (Pixel 8 Pro Emulator)
- op-sqlite version: 6.0.7
- RN version: 0.74.2
- New Architecture: Yes

**Reproducible example**

```js
async function run() {
  const db = open({ name: 'test' })
  db.execute(
    'CREATE TABLE IF NOT EXISTS test (id INTEGER PRIMARY KEY, name TEXT)'
  )

  await db.transaction(async (tx) => {
    await Promise.all(
      Array.from({ length: 100 }, (_, i) =>
        tx
          .executeAsync('INSERT INTO test (name) VALUES (?)', ['foo'])
          .then((result) => console.log('insertId:', result.insertId))
      )
    )
  })

  db.execute('DROP TABLE IF EXISTS test')
}
```

It gets:

```
 LOG  insertId: 1
 LOG  insertId: 3
 LOG  insertId: 4
 LOG  insertId: 4
 LOG  insertId: 5
 LOG  insertId: 6
 LOG  insertId: 8
 LOG  insertId: 8
 LOG  insertId: 9
 LOG  insertId: 14
 LOG  insertId: 17
 LOG  insertId: 17
 LOG  insertId: 17
 LOG  insertId: 17
 LOG  insertId: 16
 LOG  insertId: 17
 LOG  insertId: 17
 LOG  insertId: 22
 LOG  insertId: 23
 LOG  insertId: 24
 LOG  insertId: 20
 LOG  insertId: 27
 LOG  insertId: 27
 LOG  insertId: 27
 LOG  insertId: 29
 LOG  insertId: 30
 LOG  insertId: 32
 LOG  insertId: 30
 LOG  insertId: 37
 LOG  insertId: 38
 LOG  insertId: 34
 LOG  insertId: 39
 LOG  insertId: 27
 LOG  insertId: 37
 LOG  insertId: 40
 LOG  insertId: 39
 LOG  insertId: 32
 LOG  insertId: 39
 LOG  insertId: 41
 LOG  insertId: 44
 LOG  insertId: 46
 LOG  insertId: 46
 LOG  insertId: 47
 LOG  insertId: 44
 LOG  insertId: 50
 LOG  insertId: 51
 LOG  insertId: 41
 LOG  insertId: 49
 LOG  insertId: 52
 LOG  insertId: 53
 LOG  insertId: 35
 LOG  insertId: 57
 LOG  insertId: 52
 LOG  insertId: 61
 LOG  insertId: 56
 LOG  insertId: 62
 LOG  insertId: 62
 LOG  insertId: 54
 LOG  insertId: 61
 LOG  insertId: 56
 LOG  insertId: 66
 LOG  insertId: 64
 LOG  insertId: 64
 LOG  insertId: 63
 LOG  insertId: 68
 LOG  insertId: 68
 LOG  insertId: 71
 LOG  insertId: 73
 LOG  insertId: 72
 LOG  insertId: 75
 LOG  insertId: 73
 LOG  insertId: 75
 LOG  insertId: 76
 LOG  insertId: 78
 LOG  insertId: 74
 LOG  insertId: 80
 LOG  insertId: 76
 LOG  insertId: 79
 LOG  insertId: 82
 LOG  insertId: 84
 LOG  insertId: 81
 LOG  insertId: 86
 LOG  insertId: 75
 LOG  insertId: 83
 LOG  insertId: 80
 LOG  insertId: 87
 LOG  insertId: 85
 LOG  insertId: 88
 LOG  insertId: 89
 LOG  insertId: 90
 LOG  insertId: 91
 LOG  insertId: 92
 LOG  insertId: 93
 LOG  insertId: 94
 LOG  insertId: 95
 LOG  insertId: 96
 LOG  insertId: 97
 LOG  insertId: 98
 LOG  insertId: 99
 LOG  insertId: 100
```

**Comment by craftzdog** [comment]
I didn't specify the performance mode. If I specify the mode to 2 like so:

```json
  "op-sqlite": {
    "performanceMode": "2",
    "iosSqlite": false,
    "fts5": true
  },
```

but the issue still persists.
Yes, all the data looks correctly written.

**Comment by ospfranco** [comment]
Ok, I will take a look when possible

**Comment by ospfranco** [comment]
Just tried your code on the example app and I don't see any repeated insertions.

```
 (NOBRIDGE) LOG  insertId: 1
 (NOBRIDGE) LOG  insertId: 3
 (NOBRIDGE) LOG  insertId: 5
 (NOBRIDGE) LOG  insertId: 6
 (NOBRIDGE) LOG  insertId: 7
 (NOBRIDGE) LOG  insertId: 9
 (NOBRIDGE) LOG  insertId: 9
 (NOBRIDGE) LOG  insertId: 10
 (NOBRIDGE) LOG  insertId: 3
 (NOBRIDGE) LOG  insertId: 11
 (NOBRIDGE) LOG  insertId: 4
 (NOBRIDGE) LOG  insertId: 14
 (NOBRIDGE) LOG  insertId: 16
 (NOBRIDGE) LOG  insertId: 12
 (NOBRIDGE) LOG  insertId: 13
 (NOBRIDGE) LOG  insertId: 20
 (NOBRIDGE) LOG  insertId: 15
 (NOBRIDGE) LOG  insertId: 21
 (NOBRIDGE) LOG  insertId: 18
 (NOBRIDGE) LOG  insertId: 19
 (NOBRIDGE) LOG  insertId: 17
 (NOBRIDGE) LOG  insertId: 23
 (NOBRIDGE) LOG  insertId: 27
 (NOBRIDGE) LOG  insertId: 28
 (NOBRIDGE) LOG  insertId: 29
 (NOBRIDGE) LOG  insertId: 30
 (NOBRIDGE) LOG  insertId: 22
 (NOBRIDGE) LOG  insertId: 26
 (NOBRIDGE) LOG  insertId: 24
 (NOBRIDGE) LOG  insertId: 31
 (NOBRIDGE) LOG  insertId: 25
 (NOBRIDGE) LOG  insertId: 33
 (NOBRIDGE) LOG  insertId: 36
 (NOBRIDGE) LOG  insertId: 35
 (NOBRIDGE) LOG  insertId: 32
 (NOBRIDGE) LOG  insertId: 38
 (NOBRIDGE) LOG  insertId: 41
 (NOBRIDGE) LOG  insertId: 34
 (NOBRIDGE) LOG  insertId: 43
 (NOBRIDGE) LOG  insertId: 40
 (NOBRIDGE) LOG  insertId: 37
 (NOBRIDGE) LOG  insertId: 44
 (NOBRIDGE) LOG  insertId: 42
 (NOBRIDGE) LOG  insertId: 39
 (NOBRIDGE) LOG  insertId: 47
 (NOBRIDGE) LOG  insertId: 46
 (NOBRIDGE) LOG  insertId: 48
 (NOBRIDGE) LOG  insertId: 49
 (NOBRIDGE) LOG  insertId: 50
 (NOBRIDGE) LOG  insertId: 46
 (NOBRIDGE) LOG  insertId: 51
 (NOBRIDGE) LOG  insertId: 52
 (NOBRIDGE) LOG  insertId: 53
 (NOBRIDGE) LOG  insertId: 55
 (NOBRIDGE) LOG  insertId: 54
 (NOBRIDGE) LOG  insertId: 56
 (NOBRIDGE) LOG  insertId: 58
 (NOBRIDGE) LOG  insertId: 60
 (NOBRIDGE) LOG  insertId: 62
 (NOBRIDGE) LOG  insertId: 61
 (NOBRIDGE) LOG  insertId: 60
 (NOBRIDGE) LOG  insertId: 65
 (NOBRIDGE) LOG  insertId: 64
 (NOBRIDGE) LOG  insertId: 61
 (NOBRIDGE) LOG  insertId: 67
 (NOBRIDGE) LOG  insertId: 66
 (NOBRIDGE) LOG  insertId: 68
 (NOBRIDGE) LOG  insertId: 71
 (NOBRIDGE) LOG  insertId: 70
 (NOBRIDGE) LOG  insertId: 73
 (NOBRIDGE) LOG  insertId: 69
 (NOBRIDGE) LOG  insertId: 63
 (NOBRIDGE) LOG  insertId: 74
 (NOBRIDGE) LOG  insertId: 72
 (NOBRIDGE) LOG  insertId: 76
 (NOBRIDGE) LOG  insertId: 75
 (NOBRIDGE) LOG  insertId: 78
 (NOBRIDGE) LOG  insertId: 79
 (NOBRIDGE) LOG  insertId: 77
 (NOBRIDGE) LOG  insertId: 80
 (NOBRIDGE) LOG  insertId: 81
 (NOBRIDGE) LOG  insertId: 83
 (NOBRIDGE) LOG  insertId: 84
 (NOBRIDGE) LOG  insertId: 85
 (NOBRIDGE) LOG  insertId: 82
 (NOBRIDGE) LOG  insertId: 89
 (NOBRIDGE) LOG  insertId: 87
 (NOBRIDGE) LOG  insertId: 88
 (NOBRIDGE) LOG  insertId: 90
 (NOBRIDGE) LOG  insertId: 86
 (NOBRIDGE) LOG  insertId: 92
 (NOBRIDGE) LOG  insertId: 96
 (NOBRIDGE) LOG  insertId: 94
 (NOBRIDGE) LOG  insertId: 98
 (NOBRIDGE) LOG  insertId: 91
 (NOBRIDGE) LOG  insertId: 95
 (NOBRIDGE) LOG  insertId: 93
 (NOBRIDGE) LOG  insertId: 100
 (NOBRIDGE) LOG  insertId: 97
 (NOBRIDGE) LOG  insertId: 99
 ```
 
 Any more info?

**Comment by ospfranco** [comment]
There might be subtle race condition changes between the new arch and the old arch. I think the behavior is correct. The problem is sqlite has only global pointers for errors, insertion ids and other values that are set after each statement execution.

`Promise.all` is IMO an anti pattern as it offers no control about the order of execution of the operations. You are creating 100 promises which will all start being queued/dequeued from the event loop, so at some point there is a small period of time where C++ is not fast enough to [read the last insertion id from the global register](https://github.com/OP-Engineering/op-sqlite/blob/main/cpp/bridge.cpp#L533) before it gets overwritten by the next operation being executed.

As all your statements are being executed in the same table and mutexes are not implemented on the C++ level but only on the transaction level, you should rather do:

```ts
async myFn() {
	await db.transaction(async (tx) => {
		...
		for (const num of myArr) {
			let res = await tx.executeAsync('INSERT INTO test (name) VALUES (?)', ['foo'])
			console.log('insertId:', res.insertId)
		}
	}
}
```

or an old school loop:

```ts
async myFn() {
	await db.transaction(async (tx) => {
		...
		for (let i = 0; i < 100; i++) {
			let res = await tx.executeAsync('INSERT INTO test (name) VALUES (?)', ['foo'])
			console.log('insertId:', res.insertId)
		}
	}
}
```

The only other alternative would be to re-rewrite the C++ side of things to have a far smarter locking mechanism. But it gets complex fast. One does not want to lock when doing operations on different tables, also not on read operations, also not... etc etc.

**Comment by craftzdog** [comment]
Thanks for looking into it! I understand that.
I happened to encounter this issue when making a PouchDB websql adapter directly use op-sqlite.
The implementation assumes that every writing SQL request is executed sequentially in websql.
So, there is a callback function for `bulkDocs` that gets invoked a lot of times at the same JS event loop just like `Promise.all`.
The adapter uses `insertId` as its change history IDs in the database.
I managed to mitigate it by making a simple queue as you suggested.

Regarding the locking mechanism, [`sqlite3_stmt_readonly`](https://www.sqlite.org/c3ref/stmt_readonly.html) might be useful to detect if the statement makes any changes to the table.
So, it'd be nice to have a critical section only for writing operations, perhaps?

Thanks again for creating the awesome library!
